src/state.ts

```diff
@@ -24,62 +24,129 @@ export type TransitionEvent<TContext = unknown, TEvent = unknown> = (
 	hookInput: HookInput<TContext, TEvent>,
 ) => boolean | Promise<boolean>;
 
-export type GenericInitialState<TContext = unknown, TEvent = unknown> =
-	| InitialGuardState
-	| InitialAutoTransitionState;
-
-export interface InitialGuardState<TContext = unknown, TEvent = unknown> {
+export interface InitialState<TContext = unknown, TEvent = unknown> {
 	id: StateIdentifier;
-	initial: boolean;
+	initial: true;
+	transitionTo: TransitionToHook<TContext, TEvent>;
+
+	autoTransition?: false;
 	transitionGuard?: TransitionEvent<TContext, TEvent>;
+}
+
+export interface InitialWithFinalFalseState<
+	TContext = unknown,
+	TEvent = unknown,
+> {
+	id: StateIdentifier;
+	initial: true;
 	transitionTo: TransitionToHook<TContext, TEvent>;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
+	final: false;
+
+	autoTransition?: false;
+	transitionGuard?: TransitionEvent<TContext, TEvent>;
 }
 
-export interface InitialAutoTransitionState<
+export interface InitialWithAutoTransitionState<
 	TContext = unknown,
 	TEvent = unknown,
 > {
 	id: StateIdentifier;
-	initial: boolean;
-	autoTransition: boolean;
+	initial: true;
+	autoTransition: true;
 	transitionTo: TransitionToHook<TContext, TEvent>;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
 }
 
-export interface FinalState<TContext = unknown, TEvent = unknown> {
+export interface InitialWithAutoTransitionAndFinalFalseState<
+	TContext = unknown,
+	TEvent = unknown,
+> {
 	id: StateIdentifier;
-	final: boolean;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
-	onFinal?: OnFinalHook<TContext>;
+	initial: true;
+	autoTransition: true;
+	transitionTo: TransitionToHook<TContext, TEvent>;
+	final: false;
 }
 
-export type GenericState<TContext = unknown, TEvent = unknown> =
-	| GuardState
-	| AutoTransitionState;
+export interface NonFinalState<TContext = unknown, TEvent = unknown> {
+	id: StateIdentifier;
+	transitionTo: TransitionToHook<TContext, TEvent>;
 
-export interface GuardState<TContext = unknown, TEvent = unknown> {
+	initial?: boolean;
+	autoTransition?: false;
+	transitionGuard?: TransitionEvent<TContext, TEvent>;
+}
+
+export interface NonFinalWithFinalFalseState<
+	TContext = unknown,
+	TEvent = unknown,
+> {
 	id: StateIdentifier;
+	transitionTo: TransitionToHook<TContext, TEvent>;
+	final: false;
+
+	initial?: boolean;
+	autoTransition?: false;
 	transitionGuard?: TransitionEvent<TContext, TEvent>;
+}
+
+export interface NonFinalWithAutoTransitionState<
+	TContext = unknown,
+	TEvent = unknown,
+> {
+	id: StateIdentifier;
+	autoTransition: true;
 	transitionTo: TransitionToHook<TContext, TEvent>;
-	onEntry?: OnEntryHook<TContext, TEvent>;
-	onExit?: OnExitHook<TContext, TEvent>;
+
+	initial?: boolean;
 }
 
-export interface AutoTransitionState<TContext = unknown, TEvent = unknown> {
+export interface NonFinalWithAutoTransitionAndFinalFalseState<
```

Comment:
there are 3 kind of state:
`initial`, `final`, and one that is non initial and non final.
The name `NonFinal` doesnt make sense because it includes initial states too